src/main/java/com/flowpowered/nbt/regionfile/RegionFile.java

```diff
@@ -0,0 +1,235 @@
+ * which will be loaded completely in the constructor. Warning: typical Minecraft region files may take up to several MiB of space and the
+ * theoretical maximum size is at one GiB (1024 region files take up at most 255 sectors à 4096 bytes). <br/>
+ * Once the object has been created, it will close the opened {@link FileChannel}. No data will be read anymore. It will be assumed that the
+ * file will not be modified externally in any ways. All changes to the objects are kept in memory and only written back to the file upon
+ * calling {@link #writeChanges()}, in which case a new file channel is opened.
+ *
+ * @author piegames
+ */
+public class RegionFile implements Iterable<Chunk> {
+
+	protected final Path	file;
+
+	protected final Chunk[]	chunks;
+	protected boolean[]		chunksUpToDate;
+
+	protected ByteBuffer	locations;
+	protected IntBuffer		locations2;
+	protected ByteBuffer	timestamps;
+	protected IntBuffer		timestamps2;
+	protected ShortBuffer	sectorCounts;
+	protected List<Boolean>	sectorUsed;
+
+	/**
+	 * Create a new RegionFile object representing the region file at the given path and load it into memory.
+	 * 
+	 * @throws IllegalArgumentException
+	 *             if the file is smaller than 4kiB
+	 * @throws NoSuchFileException
+	 *             if the file does not exist
+	 * @see #open(Path)
+	 * @author piegames
+	 */
+	public RegionFile(Path file) throws IOException {
+		this.file = Objects.requireNonNull(file);
+
+		if (!Files.exists(file))
+			throw new NoSuchFileException(file.toString());
+		if (Files.size(file) < 4096 * 2)
+			throw new IllegalArgumentException("File size must be at least 4kiB, is this file corrupt?");
+		FileChannel raf = FileChannel.open(file, StandardOpenOption.READ);
+
+		locations = ByteBuffer.allocate(4096);
+		raf.read(locations);
+		locations.flip();
+		locations2 = locations.asIntBuffer();
+
+		timestamps = ByteBuffer.allocate(4096);
+		raf.read(timestamps);
+		timestamps.flip();
+		timestamps2 = timestamps.asIntBuffer();
+
+		chunks = new Chunk[32 * 32];
+		chunksUpToDate = new boolean[32 * 32];
+		for (int i = 0; i < 1024; i++) {
+			int chunkPos = locations2.get(i) >>> 8;
+			int chunkLength = locations2.get(i) & 0xFF;
+			if (chunkPos > 0) {
+				/* i & 31 retrieves the last 5 bit which store the x coordinate */
+				chunks[i] = new Chunk(i & 31, i >> 5, timestamps2.get(i), raf, chunkPos, chunkLength);
+			} else
+				chunks[i] = null;
+			chunksUpToDate[i] = true;
+		}
+	}
+
+	/**
+	 * Returns the {@link Chunk} at the given coordinate
+	 *
+	 * @param x
+	 *            the x coordinate relative to this region file's origin
+	 * @param z
+	 *            the z coordinate relative to this region file's origin
+	 * @return the chunk at that coordinate or {@code null} if the chunk does not exist
+	 * @author piegames
+	 */
+	public Chunk getChunk(int x, int z) {
```

Comment:
This method should return Optional<Chunk> instead of a nullable Chunk.
At the vary least annotate this method with `@Nullable`

Reply:
I'd like to, but if I understand it correctly we'd need to add a library to the class path to get those annotations.